File: golang/pkg/agentbay/oss/oss.go
package oss

import (
	"fmt"

	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// EnvInitResult represents the result of OSS environment initialization
type EnvInitResult struct {
	models.ApiResponse
	Result string
}

// UploadResult represents the result of a file upload operation
type UploadResult struct {
	models.ApiResponse
	URL string
}

// DownloadResult represents the result of a file download operation
type DownloadResult struct {
	models.ApiResponse
	LocalPath string
}

// OSSManager handles Object Storage Service operations in the AgentBay cloud environment
type OSSManager struct {
	Session interface {
		GetAPIKey() string
		GetClient() *mcp.Client
		GetSessionId() string
		IsVpc() bool
		NetworkInterfaceIp() string
		HttpPort() string
		FindServerForTool(toolName string) string
		CallMcpTool(toolName string, args interface{}, autoGenSession ...bool) (*models.McpToolResult, error)
	}
}

// NewOss creates a new OSS manager instance
func NewOss(session interface {
	GetAPIKey() string
	GetClient() *mcp.Client
	GetSessionId() string
	IsVpc() bool
	NetworkInterfaceIp() string
	HttpPort() string
	FindServerForTool(toolName string) string
	CallMcpTool(toolName string, args interface{}, autoGenSession ...bool) (*models.McpToolResult, error)
}) *OSSManager {
	return &OSSManager{
		Session: session,
	}
}

// EnvInit initializes OSS environment variables with STS temporary credentials.
// All three credential parameters (accessKeyId, accessKeySecret, securityToken) are required for security.
//
// Example:
//
//	client, _ := agentbay.NewAgentBay(os.Getenv("AGENTBAY_API_KEY"), nil)
//	result, _ := client.Create(nil)
//	defer result.Session.Delete()
//	ossResult, _ := result.Session.Oss.EnvInit("stsAccessKeyId", "stsAccessKeySecret", "stsToken", "endpoint", "cn-hangzhou")
func (o *OSSManager) EnvInit(accessKeyId, accessKeySecret, securityToken, endpoint, region string) (*EnvInitResult, error) {
	// Build arguments map
	args := map[string]interface{}{
		"access_key_id":     accessKeyId,
		"access_key_secret": accessKeySecret,
		"security_token":    securityToken,
	}

	// Add optional parameters if provided
	if endpoint != "" {
		args["endpoint"] = endpoint
	}
	if region != "" {
		args["region"] = region
	}

	result, err := o.Session.CallMcpTool("oss_env_init", args)
	if err != nil {
		return nil, fmt.Errorf("error initializing OSS environment: %w", err)
	}

	return &EnvInitResult{
		ApiResponse: models.ApiResponse{
			RequestID: result.RequestID,
		},
		Result: result.Data,
	}, nil
}

// Upload uploads a local file or directory to the specified OSS bucket
//
// Note: Before calling this API, you must call EnvInit to initialize the OSS environment.
//
// Example:
//
//	client, _ := agentbay.NewAgentBay(os.Getenv("AGENTBAY_API_KEY"), nil)
//	result, _ := client.Create(nil)
//	defer result.Session.Delete()
//	result.Session.Oss.EnvInit("accessKeyId", "accessKeySecret", "token", "endpoint", "cn-hangzhou")
//	uploadResult, _ := result.Session.Oss.Upload("my-bucket", "my-object", "/tmp/file.txt")
func (o *OSSManager) Upload(bucket, object, path string) (*UploadResult, error) {
	args := map[string]interface{}{
		"bucket": bucket,
		"object": object,
		"path":   path,
	}

	result, err := o.Session.CallMcpTool("oss_upload", args)
	if err != nil {
		return nil, fmt.Errorf("error uploading to OSS: %w", err)
	}

	return &UploadResult{
		ApiResponse: models.ApiResponse{
			RequestID: result.RequestID,
		},
		URL: result.Data,
	}, nil
}

// UploadAnonymous uploads a local file or directory to the specified URL using HTTP PUT
//
// Note: Before calling this API, you must call EnvInit to initialize the OSS environment.
//
// Example:
//
//	client, _ := agentbay.NewAgentBay(os.Getenv("AGENTBAY_API_KEY"), nil)
//	result, _ := client.Create(nil)
//	defer result.Session.Delete()
//	result.Session.Oss.EnvInit("accessKeyId", "accessKeySecret", "token", "endpoint", "cn-hangzhou")
//	uploadResult, _ := result.Session.Oss.UploadAnonymous("https://example.com/upload", "/tmp/file.txt")
func (o *OSSManager) UploadAnonymous(url, path string) (*UploadResult, error) {
	args := map[string]interface{}{
		"url":  url,
		"path": path,
	}

	result, err := o.Session.CallMcpTool("oss_upload_annon", args)
	if err != nil {
		return nil, fmt.Errorf("error uploading anonymously: %w", err)
	}

	return &UploadResult{
		ApiResponse: models.ApiResponse{
			RequestID: result.RequestID,
		},
		URL: result.Data,
	}, nil
}

// Download downloads an object from the specified OSS bucket to the given local path
//
// Note: Before calling this API, you must call EnvInit to initialize the OSS environment.
//
// Example:
//
//	client, _ := agentbay.NewAgentBay(os.Getenv("AGENTBAY_API_KEY"), nil)
//	result, _ := client.Create(nil)
//	defer result.Session.Delete()
//	result.Session.Oss.EnvInit("accessKeyId", "accessKeySecret", "token", "endpoint", "cn-hangzhou")
//	downloadResult, _ := result.Session.Oss.Download("my-bucket", "my-object", "/tmp/file.txt")
func (o *OSSManager) Download(bucket, object, path string) (*DownloadResult, error) {
	args := map[string]interface{}{
		"bucket": bucket,
		"object": object,
		"path":   path,
	}

	result, err := o.Session.CallMcpTool("oss_download", args)
	if err != nil {
		return nil, fmt.Errorf("error downloading from OSS: %w", err)
	}

	return &DownloadResult{
		ApiResponse: models.ApiResponse{
			RequestID: result.RequestID,
		},
		LocalPath: path,
	}, nil
}

// DownloadAnonymous downloads a file from the specified URL to the given local path
//
// Note: Before calling this API, you must call EnvInit to initialize the OSS environment.
//
// Example:
//
//	client, _ := agentbay.NewAgentBay(os.Getenv("AGENTBAY_API_KEY"), nil)
//	result, _ := client.Create(nil)
//	defer result.Session.Delete()
//	result.Session.Oss.EnvInit("accessKeyId", "accessKeySecret", "token", "endpoint", "cn-hangzhou")
//	downloadResult, _ := result.Session.Oss.DownloadAnonymous("https://example.com/file.txt", "/tmp/file.txt")
func (o *OSSManager) DownloadAnonymous(url, path string) (*DownloadResult, error) {
	args := map[string]interface{}{
		"url":  url,
		"path": path,
	}

	result, err := o.Session.CallMcpTool("oss_download_annon", args)
	if err != nil {
		return nil, fmt.Errorf("error downloading anonymously: %w", err)
	}

	return &DownloadResult{
		ApiResponse: models.ApiResponse{
			RequestID: result.RequestID,
		},
		LocalPath: path,
	}, nil
}

File: python/agentbay/oss/oss.py
import json
from typing import Any, Dict, Optional

from agentbay.api.base_service import BaseService
from ..logger import get_logger, _log_api_response
from agentbay.exceptions import AgentBayError, OssError
from agentbay.model import ApiResponse

# Initialize logger for this module
_logger = get_logger("oss")


class OSSClientResult(ApiResponse):
    """Result of OSS client creation operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        client_config: Optional[Dict[str, Any]] = None,
        error_message: str = "",
    ):
        """
        Initialize an OSSClientResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            client_config (Dict[str, Any], optional): OSS client configuration.
                Defaults to None.
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.client_config = client_config or {}
        self.error_message = error_message


class OSSUploadResult(ApiResponse):
    """Result of OSS upload operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        content: str = "",
        error_message: str = "",
    ):
        """
        Initialize an OSSUploadResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            content (str, optional): Result of the upload operation. Defaults to "".
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.content = content
        self.error_message = error_message


class OSSDownloadResult(ApiResponse):
    """Result of OSS download operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        content: str = "",
        error_message: str = "",
    ):
        """
        Initialize an OSSDownloadResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            content (string, optional): Defaults to "Download success"
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.content = content
        self.error_message = error_message


class Oss(BaseService):
    """
    Handles Object Storage Service operations in the AgentBay cloud environment.
    """

    def __init__(self, session):
        """
        Initialize an Oss object.

        Args:
            session: The Session instance that this Oss belongs to.
        """
        super().__init__(session)

    def _handle_error(self, e):
        """
        Convert AgentBayError to OssError for compatibility.

        Args:
            e (Exception): The exception to convert.

        Returns:
            OssError: The converted exception.
        """
        if isinstance(e, OssError):
            return e
        if isinstance(e, AgentBayError):
            return OssError(str(e))
        return e

    def env_init(
        self,
        access_key_id: str,
        access_key_secret: str,
        securityToken: str,
        endpoint: Optional[str] = None,
        region: Optional[str] = None,
    ) -> OSSClientResult:
        """
        Create an OSS client with the provided STS temporary credentials.

        Args:
            access_key_id: The Access Key ID from STS temporary credentials.
            access_key_secret: The Access Key Secret from STS temporary credentials.
            securityToken: Security token from STS temporary credentials. Required for security.
            endpoint: The OSS service endpoint. If not specified, the default is used.
            region: The OSS region. If not specified, the default is used.

        Returns:
            OSSClientResult: Result object containing client configuration and error
                message if any.

        Example:
            ```python
            session = agent_bay.create().session
            session.oss.env_init(
                access_key_id="your_sts_access_key_id",
                access_key_secret="your_sts_access_key_secret",
                securityToken="your_sts_security_token"
            )
            session.delete()
            ```
        """
        try:
            args = {
                "access_key_id": access_key_id,
                "access_key_secret": access_key_secret,
                "security_token": securityToken,
            }

            # Add optional parameters if provided
            if endpoint:
                args["endpoint"] = endpoint
            if region:
                args["region"] = region

            result = self.session.call_mcp_tool("oss_env_init", args)
            try:
                response_body = json.dumps(
                    getattr(result, "body", result), ensure_ascii=False, indent=2
                )
                _log_api_response(response_body)
            except Exception:
                _logger.debug(f"📥 Response: {result}")

            if result.success:
                try:
                    client_config = result.data
                    return OSSClientResult(
                        request_id=result.request_id,
                        success=True,
                        client_config=client_config,
                    )
                except json.JSONDecodeError:
                    return OSSClientResult(
                        request_id=result.request_id,
                        success=False,
                        error_message="Failed to parse client configuration JSON",
                    )
            else:
                return OSSClientResult(
                    request_id=result.request_id,
                    success=False,
                    error_message=result.error_message
                    or "Failed to initialize OSS environment",
                )
        except AgentBayError as e:
            return OSSClientResult(request_id="", success=False, error_message=str(e))
        except Exception as e:
            return OSSClientResult(
                request_id="",
                success=False,
                error_message=f"Failed to initialize OSS environment: {e}",
            )

    def upload(self, bucket: str, object: str, path: str) -> OSSUploadResult:
        """
        Upload a local file or directory to OSS.

        Note: Before calling this API, you must first call env_init to initialize
            the OSS environment.

        Args:
            bucket: OSS bucket name.
            object: Object key in OSS.
            path: Local file or directory path to upload.

        Returns:
            OSSUploadResult: Result object containing upload result and error message
                if any.

        Example:
            ```python
            session = agent_bay.create().session
            session.oss.env_init(
                access_key_id="your_access_key_id",
                access_key_secret="your_access_key_secret"
            )
            result = session.oss.upload("my-bucket", "file.txt", "/local/path/file.txt")
            print(f"Upload result: {result.content}")
            session.delete()
            ```
        """
        try:
            args = {"bucket": bucket, "object": object, "path": path}

            result = self.session.call_mcp_tool("oss_upload", args)
            _logger.debug(f"📥 OSS Response: {result}")

            if result.success:
                return OSSUploadResult(
                    request_id=result.request_id,
                    success=True,
                    content=result.data,
                )
            else:
                return OSSUploadResult(
                    request_id=result.request_id,
                    success=False,
                    error_message=result.error_message or "Failed to upload to OSS",
                )
        except AgentBayError as e:
            return OSSUploadResult(request_id="", success=False, error_message=str(e))
        except Exception as e:
            return OSSUploadResult(
                request_id="",
                success=False,
                error_message=f"Failed to upload to OSS: {e}",
            )

    def upload_anonymous(self, url: str, path: str) -> OSSUploadResult:
        """
        Upload a local file or directory to a URL anonymously.

        Args:
            url: The HTTP/HTTPS URL to upload the file to.
            path: Local file or directory path to upload.

        Returns:
            OSSUploadResult: Result object containing upload result and error message
                if any.

        Example:
            ```python
            session = agent_bay.create().session
            result = session.oss.upload_anonymous(
                "https://example.com/upload",
                "/local/path/file.txt"
            )
            print(f"Upload result: {result.content}")
            session.delete()
            ```
        """
        try:
            args = {"url": url, "path": path}

            result = self.session.call_mcp_tool("oss_upload_annon", args)
            _logger.debug(f"📥 OSS Response: {result}")

            if result.success:
                return OSSUploadResult(
                    request_id=result.request_id,
                    success=True,
                    content=result.data,
                )
            else:
                return OSSUploadResult(
                    request_id=result.request_id,
                    success=False,
                    error_message=result.error_message
                    or "Failed to upload anonymously",
                )
        except AgentBayError as e:
            return OSSUploadResult(request_id="", success=False, error_message=str(e))
        except Exception as e:
            return OSSUploadResult(
                request_id="",
                success=False,
                error_message=f"Failed to upload anonymously: {e}",
            )

    def download(self, bucket: str, object: str, path: str) -> OSSDownloadResult:
        """
        Download an object from OSS to a local file or directory.

        Note: Before calling this API, you must first call env_init to initialize
            the OSS environment.

        Args:
            bucket: OSS bucket name.
            object: Object key in OSS.
            path: Local file or directory path to download to.

        Returns:
            OSSDownloadResult: Result object containing download status and error
                message if any.

        Example:
            ```python
            session = agent_bay.create().session
            session.oss.env_init(
                access_key_id="your_access_key_id",
                access_key_secret="your_access_key_secret"
            )
            result = session.oss.download("my-bucket", "file.txt", "/local/path/file.txt")
            print(f"Download result: {result.content}")
            session.delete()
            ```
        """
        try:
            args = {"bucket": bucket, "object": object, "path": path}

            result = self.session.call_mcp_tool("oss_download", args)
            _logger.debug(f"📥 OSS Response: {result}")

            if result.success:
                return OSSDownloadResult(
                    request_id=result.request_id,
                    success=True,
                    content=result.data,
                )
            else:
                return OSSDownloadResult(
                    request_id=result.request_id,
                    success=False,
                    error_message=result.error_message or "Failed to download from OSS",
                )
        except AgentBayError as e:
            return OSSDownloadResult(request_id="", success=False, error_message=str(e))
        except Exception as e:
            return OSSDownloadResult(
                request_id="",
                success=False,
                error_message=f"Failed to download from OSS: {e}",
            )

    def download_anonymous(self, url: str, path: str) -> OSSDownloadResult:
        """
        Download a file from a URL anonymously to a local file path.

        Args:
            url: The HTTP/HTTPS URL to download the file from.
            path: Local file or directory path to download to.

        Returns:
            OSSDownloadResult: Result object containing download status and error
                message if any.

        Example:
            ```python
            session = agent_bay.create().session
            result = session.oss.download_anonymous(
                "https://example.com/file.txt",
                "/local/path/file.txt"
            )
            print(f"Download result: {result.content}")
            session.delete()
            ```
        """
        try:
            args = {"url": url, "path": path}

            result = self.session.call_mcp_tool("oss_download_annon", args)
            _logger.debug(f"📥 OSS Response: {result}")

            if result.success:
                return OSSDownloadResult(
                    request_id=result.request_id,
                    success=True,
                    content=result.data,
                )
            else:
                return OSSDownloadResult(
                    request_id=result.request_id,
                    success=False,
                    error_message=result.error_message
                    or "Failed to download anonymously",
                )
        except AgentBayError as e:
            return OSSDownloadResult(request_id="", success=False, error_message=str(e))
        except Exception as e:
            return OSSDownloadResult(
                request_id="",
                success=False,
                error_message=f"Failed to download anonymously: {e}",
            )
